aptos-move/aptos-vm/src/data_cache.rs

```diff
@@ -107,72 +146,168 @@ impl<'a, S: StateView> StorageAdapter<'a, S> {
     ) -> Result<(Option<Vec<u8>>, usize), VMError> {
         let resource_group = get_resource_group_from_metadata(struct_tag, metadata);
         if let Some(resource_group) = resource_group {
-            let mut cache = self.resource_group_cache.borrow_mut();
-            let cache = cache.entry(*address).or_insert_with(BTreeMap::new);
-            if let Some(group_data) = cache.get_mut(&resource_group) {
-                // This resource group is already cached for this address. So just return the
-                // cached value.
-                let buf = group_data.get(struct_tag).cloned();
-                let buf_size = resource_size(&buf);
-                return Ok((buf, buf_size));
-            }
-            let group_data = self.get_resource_group_data(address, &resource_group)?;
-            if let Some(group_data) = group_data {
-                let len = if self.accurate_byte_count {
-                    group_data.len()
-                } else {
-                    0
-                };
-                let group_data: BTreeMap<StructTag, Vec<u8>> = bcs::from_bytes(&group_data)
-                    .map_err(|_| {
+            let key = StateKey::access_path(AccessPath::resource_group_access_path(
+                *address,
+                resource_group.clone(),
+            ));
+
+            let first_access = match &mut *self.resource_group_cache.borrow_mut() {
+                ResourceGroupCache::V0(ref mut btree_cache) => {
+                    if let Some(group_data) = btree_cache.get_mut(&key) {
+                        // This resource group is already V0-cached for this address.
+                        // So just return the cached value.
+                        let buf = group_data.get(struct_tag).cloned();
+                        let buf_size = resource_size(&buf);
+                        return Ok((buf, buf_size));
+                    } else {
+                        true
+                    }
+                },
+                ResourceGroupCache::V1(ref mut accesses) => accesses.insert(key.clone()),
+            };
+
+            let (buf, maybe_group_size) = self
+                .get_resource_from_group(&key, struct_tag, first_access && self.accurate_byte_count)
+                .map_err(|e| {
+                    if self.group_byte_count_as_sum {
+                        // Message also gated with the byte counting as sum flag / gas version.
```

Comment:
I am wondering if adding a message should be backwards compatible in terms of txn replay...